File: python/bowling_ball_selection.py
from collections import Counter
import math


def solve1():
    n,m = map(int,input().split())
    weights= map(int,input().split())
    total_num = math.comb(n,2)

    same_weights = Counter(weights)
    for i in range(1,m+1):
        if same_weights[i]>1:
            total_num -=math.comb(same_weights[i],2)
            
    print(total_num)

def solve2():
    n,m=map(int,input().split())
    arr = list(map(int,input().split()))
    num_count=[0]*(m+1)

    for num in arr:
        num_count[num]+=1
    
    result=0
    for i in range(1,m+1):
        cur = num_count[i]
        n -=cur
        result += cur*n
    print(result)


if __name__=="__main__":
    solve1()
    solve2()

File: c++/bowling_ball_selection.cpp
#include <vector>
#include <iostream>
#include <math.h>
using namespace std;


int factorial(int n){
    int ret=1;
    for(int i=n;i>1;i--)
        ret*=i;
    return ret;
}

int comb(int n,int r){
    return factorial(n)/(factorial(n-r)*factorial(r));
}

void solve(){
    int n,m;
    cin >> n >> m;
    vector<int>weights(n);
    vector<int>weights_num(m+1,0);
    for(int i=0;i<n;i++){
        cin >> weights[i];
    }
    int total_select = comb(n,2);
    for(int i=0;i<weights.size();i++){
        weights_num[weights[i]]++;
    }
    for(int i=1;i<=m;i++){
        if(weights_num[i]>1){
            total_select -= comb(weights_num[i],2);
        }
    }
    cout << total_select << "\n";
}

int main(){
    solve();
}